### ISSUE TITLE
[Feature Request] Add support for overlapping segments

### ISSUE THREAD

**Issue by fedorov** [OHIF/Viewers#3496]
### What feature or change would you like to see made?

Overlapping DICOM SEG segments are currently no loaded with the error message indicating they are not supported in v3. Overlapping segments were supported in v2.

### Why should we prioritize this feature?

IDC project is currently using v2, which is able to render overlapping segments. IDC contains data that contains overlapping segments, such as this one: https://viewer.imaging.datacommons.cancer.gov/viewer/1.3.6.1.4.1.32722.99.99.62087908186665265759322018723889952421. Unless this feature is implemented, this would be a regression if IDC switches to v3.

**Comment by igoroctaviano** [comment]
@sedghi @rodrigobasilio2022 will discuss a solution to this issue. 
A stack viewport is being implemented and might address this issue.

**Comment by james-hanks** [comment]
@sedghi link stack segmentation ticket (if we have one)

**Comment by fedorov** [comment]
Per @rodrigobasilio2022 (if I understand correctly), this is blocked by https://github.com/cornerstonejs/cornerstone3D/pull/679.

**Comment by rodrigobasilio2022** [comment]
Yes, the steps are:
- Finish this PR #679;
- Add overlapping segmentations in a new PR;
- Implement the modifications in OHIF;

**Comment by igoroctaviano** [comment]
Related to #3581

**Comment by james-hanks** [comment]
On hold because related fix will be published through industry partner project Q1 2024

**Comment by fedorov** [comment]
@james-hanks @igoroctaviano I observe that overlapping segments that are from different segmentations do work. Is this expected - that overlapping segments from different segmentations work, but from the same segmentation do not?

https://viewer.imaging.datacommons.cancer.gov/v3/viewer/?StudyInstanceUIDs=1.3.6.1.4.1.14519.5.2.1.6279.6001.191425307197546732281885591780

![2023-12-14_11-41-55](https://github.com/OHIF/Viewers/assets/313942/d0a2230c-6f2f-4ac1-903f-b28a898457e6)

**Comment by fedorov** [comment]
> overlapping segments from different segmentations work, but from the same segmentation do not

Yes, I confirmed the above. Interesting. The examples below can be used to demonstrate this (if you have little kids, you can ask them to play the game of "find differences" - it should be fun!):

* v2: https://viewer.imaging.datacommons.cancer.gov/viewer/1.3.6.1.4.1.32722.99.99.71961866280433925571019872464419293819

![image](https://github.com/OHIF/Viewers/assets/313942/6ce5c2e7-2d1a-4dff-aa01-8502fd3baa33)

* v3: https://viewer.imaging.datacommons.cancer.gov/v3/viewer/?StudyInstanceUIDs=1.3.6.1.4.1.32722.99.99.71961866280433925571019872464419293819

![image](https://github.com/OHIF/Viewers/assets/313942/50c56dbe-76c8-40be-84ce-7a1f207b1b56)

I do have to point out that I find it very disturbing that you show incorrect segmentation overlay by cropping overlap areas, and do not give any warning to the user about this. Segmentations are shown as if everything is normal. This is really bad, IMHO. I hope no one is looking at any overlapping segments that are used for patient care using v3!

**Comment by sedghi** [comment]
Yes, if there are separate segmentations they will work fine. 

> I find it very disturbing that you show incorrect segmentation overlay by cropping overlap areas, and do not give any warning to the user about this. Segmentations are shown as if everything is normal. This is really bad, IMHO. I hope no one is looking at any overlapping segments that are used for patient care using v3!

I thought we show a warning, but tried it now and we don't, hmmmm I will look into this, but I guess your comment is about overlapping segments in one segmentation right?

**Comment by sedghi** [comment]
I'm pretty sure I had that warning, not sure how it is not showing up. I will fix it in the next PR

**Comment by fedorov** [comment]
Just discovered this today (on OHIF demo viewer) that there is still no support for overlapping segments AND there is no any kind of warning that what you see is not what you have. In this particular case, the result is very confusing.

https://viewer.ohif.org/viewer?StudyInstanceUIDs=1.3.6.1.4.1.14519.5.2.1.3671.4754.298665348758363466150039312520

<img width="640" alt="image" src="https://github.com/user-attachments/assets/dad1db6c-a44e-4299-989f-b85732d93198">

This is how it should look in reality (shown in v2): https://viewer.imaging.datacommons.cancer.gov/viewer/1.3.6.1.4.1.14519.5.2.1.3671.4754.298665348758363466150039312520

<img width="698" alt="image" src="https://github.com/user-attachments/assets/048f2074-84aa-46ab-b5e8-f5c913461fcc">

**Comment by fedorov** [comment]
> On hold because related fix will be published through industry partner project Q1 2024

For the record, it is way past Q1 2024, and this issue is still unresolved.

**Comment by fedorov** [comment]
Is there any update on the timeline for resolving this issue? 

This missing feature is one of the biggest obstacles for us in IDC to switching to v3 as the primary viewer and deprecating v2.

**Comment by sedghi** [comment]
Currently, this isn't possible. However, we have just layed the groundwork for the solution in Cornerstone3D 2.0. Once that's complete, the effort required will likely be much less.

https://github.com/cornerstonejs/cornerstone3D/pull/1400

**Comment by fedorov** [comment]
What is very confusing is that it is apparently possible to display overlapping segments that come from different segmentations, but not if overlapping segments are in the same segmentation....

Example: https://viewer.imaging.datacommons.cancer.gov/v3/viewer/?StudyInstanceUIDs=1.2.840.113654.2.55.38309775581218325422753197576029670247

**Comment by igoroctaviano** [comment]
> What is very confusing is that it is apparently possible to display overlapping segments that come from different segmentations, but not if overlapping segments are in the same segmentation....
> 
> Example: https://viewer.imaging.datacommons.cancer.gov/v3/viewer/?StudyInstanceUIDs=1.2.840.113654.2.55.38309775581218325422753197576029670247
> 
> ![2024-10-08_16-22-57](https://private-user-images.githubusercontent.com/313942/374713760-36614d6f-bbfa-4812-86da-a22d4a660235.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0MTkzOTUsIm5iZiI6MTcyODQxOTA5NSwicGF0aCI6Ii8zMTM5NDIvMzc0NzEzNzYwLTM2NjE0ZDZmLWJiZmEtNDgxMi04NmRhLWEyMmQ0YTY2MDIzNS5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwOFQyMDI0NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YmZmNWZjZDgyNDA3NjRmOTJlMWVhYTAwNmNhYTRjZjE0MWYwMjAwMTcwNzc0ZGRjNTgwNWM1N2JhZTViYjUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.NlNhxGHvPptZRrrWGYEAEkTfiC1w2NI_CeCdA8YyGCI) [ ![2024-10-08_16-22-57](https://private-user-images.githubusercontent.com/313942/374713760-36614d6f-bbfa-4812-86da-a22d4a660235.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0MTkzOTUsIm5iZiI6MTcyODQxOTA5NSwicGF0aCI6Ii8zMTM5NDIvMzc0NzEzNzYwLTM2NjE0ZDZmLWJiZmEtNDgxMi04NmRhLWEyMmQ0YTY2MDIzNS5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwOFQyMDI0NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YmZmNWZjZDgyNDA3NjRmOTJlMWVhYTAwNmNhYTRjZjE0MWYwMjAwMTcwNzc0ZGRjNTgwNWM1N2JhZTViYjUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.NlNhxGHvPptZRrrWGYEAEkTfiC1w2NI_CeCdA8YyGCI) ](https://private-user-images.githubusercontent.com/313942/374713760-36614d6f-bbfa-4812-86da-a22d4a660235.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0MTkzOTUsIm5iZiI6MTcyODQxOTA5NSwicGF0aCI6Ii8zMTM5NDIvMzc0NzEzNzYwLTM2NjE0ZDZmLWJiZmEtNDgxMi04NmRhLWEyMmQ0YTY2MDIzNS5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwOFQyMDI0NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YmZmNWZjZDgyNDA3NjRmOTJlMWVhYTAwNmNhYTRjZjE0MWYwMjAwMTcwNzc0ZGRjNTgwNWM1N2JhZTViYjUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.NlNhxGHvPptZRrrWGYEAEkTfiC1w2NI_CeCdA8YyGCI) [ ](https://private-user-images.githubusercontent.com/313942/374713760-36614d6f-bbfa-4812-86da-a22d4a660235.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0MTkzOTUsIm5iZiI6MTcyODQxOTA5NSwicGF0aCI6Ii8zMTM5NDIvMzc0NzEzNzYwLTM2NjE0ZDZmLWJiZmEtNDgxMi04NmRhLWEyMmQ0YTY2MDIzNS5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwOFQyMDI0NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YmZmNWZjZDgyNDA3NjRmOTJlMWVhYTAwNmNhYTRjZjE0MWYwMjAwMTcwNzc0ZGRjNTgwNWM1N2JhZTViYjUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.NlNhxGHvPptZRrrWGYEAEkTfiC1w2NI_CeCdA8YyGCI)

I think every segmentation has its independent volume in CS3D so segments withing an SEG share the same volume which is the source the problem. @sedghi correct me if I'm wrong.

**Comment by sedghi** [comment]
@fedorov Yes, we can render two segmentations that have different segments that overlap, but overlapping segments as part of one segmentation is not supported. I mean, you can right now detect that they're overlapping segments and create two or three non-overlapping segmentations and then render them all. I guess that would suffice for you to migrate to version 3, until i fix the overlapping segments inside one segmentation and letting brush to perform on them

**Comment by fedorov** [comment]
> you can right now detect that they're overlapping segments and create two or three non-overlapping segmentations and then render them all

It's even easier than this - there's an attribute that can be used to indicate whether segments overlap: https://dicom.innolitics.com/ciods/segmentation/segmentation-image/00620013.

In the general case, one would need to double-check, but in IDC we can potentially curate our segmentations to make sure this attribute is accurate for our data.

I guess the question is what is the timeline you think for your original approach? This will help me decide if we should invest into adding the workarounds within the current implementation (and by doing this save on the ongoing effort maintaining v2), or wait for the proper solution.

**Comment by sedghi** [comment]
Unfortunately, I don't think it will happen anytime soon. If it's easy, I can demonstrate how to manage two segmentations for rendering purposes. so we can upgrade IDC to v3.

@igoroctaviano

**Comment by fedorov** [comment]
This is very helpful. Thank you for your advice and transparency on this! We will discuss with Igor to plan the next steps.

**Comment by fedorov** [comment]
> I mean, you can right now detect that they're overlapping segments and create two or three non-overlapping segmentations and then render them all.

Would you welcome such a workaround in v3, or would they need to go into IDC fork?

**Comment by sedghi** [comment]
It is welcomed indeed

**Comment by fedorov** [comment]
As of today, per @pedrokohler, we need to wait for the release of cornerstone3d 2.0, which should be released tomorrow or on Thursday to start working.

**Comment by pedrokohler** [comment]
@fedorov FYI the first step for a solution has just been merged here https://github.com/cornerstonejs/cornerstone3D/pull/1692

We still have a final step to take and allow the overlapping segs. This work should be done in the next days (maybe couple of weeks, because of the holidays)

**Comment by pedrokohler** [comment]
@fedorov FYI I've opened the PR for the second step of the solution here https://github.com/cornerstonejs/cornerstone3D/pull/1815

As soon as it's approved, the last step is to make OHIF use the new code with the solution.

**Comment by fedorov** [comment]
@pedrokohler, as I understand from the discussion with @sedghi, he is waiting for you.

**Comment by fedorov** [comment]
https://github.com/cornerstonejs/cornerstone3D/pull/1815 was merged yesterday, thanks to the great work from @pedrokohler and @sedghi!

Pedro, Alireza - can you please add a note here with the summary of the next steps and the estimated timeline?

**Comment by pedrokohler** [comment]
@fedorov the feature has been merged on OHIF master.

**Comment by fedorov** [comment]
@pedrokohler can you reference the corresponding PR here for completeness?

**Comment by pedrokohler** [comment]
https://github.com/OHIF/Viewers/pull/4849

**Comment by sedghi** [comment]
Can we close this?

**Comment by fedorov** [comment]
Yes! It seems to work well. Nice work @pedrokohler and everyone else who contributed to this!

<img width="838" alt="Image" src="https://github.com/user-attachments/assets/796d143f-bac3-4086-87da-c1efa5bbacb1" />

<img width="830" alt="Image" src="https://github.com/user-attachments/assets/514a0280-248c-463c-b72a-910a502e4356" />
